packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/handler.ts

```diff
@@ -0,0 +1,222 @@
+    this.physicalResourceId = event.PhysicalResourceId;
+
+    const roleToAssume = event.ResourceProperties.AssumeRoleArn;
+    if (!roleToAssume) {
+      throw new Error(`AssumeRoleArn must be provided`);
+    }
+
+    eks.configureAssumeRole({
+      RoleArn: roleToAssume,
+      RoleSessionName: `AWSCDK.EKSCluster.${event.RequestType}.${this.requestId}`
+    });
+  }
+
+  public get clusterName() {
+    if (!this.physicalResourceId) {
+      throw new Error(`Cannot determie cluster name without physical resource ID`);
+    }
+
+    return this.physicalResourceId;
+  }
+
+  // ------
+  // CREATE
+  // ------
+
+  public async onCreate(): Promise<OnEventResponse> {
+    console.log('createCluster:', JSON.stringify(this.newProps, undefined, 2));
+    if (!this.newProps.roleArn) {
+      throw new Error('"roleArn" is required');
+    }
+
+    const clusterName = this.newProps.name || `${this.logicalResourceId}-${this.requestId}`;
+
+    const resp = await this.eks.createCluster({
+      name: clusterName,
+      ...this.newProps,
+    });
+
+    return {
+      PhysicalResourceId: resp.cluster!.name
+    };
+  }
+
+  public async isCreateComplete() {
+    return this.isActive();
+  }
+
+  // ------
+  // DELETE
+  // ------
+
+  public async onDelete(): Promise<OnEventResponse> {
+    console.log(`deleting cluster ${this.clusterName}`);
+    try {
+      await this.eks.deleteCluster({ name: this.clusterName });
+    } catch (e) {
+      if (e.code !== 'ResourceNotFoundException') {
+        throw e;
+      } else {
+        console.log(`cluster ${this.clusterName} not found, idempotently succeeded`);
+      }
+    }
+    return {
+      PhysicalResourceId: this.clusterName
+    };
+  }
+
+  public async isDeleteComplete(): Promise<IsCompleteResponse> {
+    console.log(`waiting for cluster ${this.clusterName} to be deleted`);
+
+    try {
+      const resp = await this.eks.describeCluster({ name: this.clusterName });
+      console.log('describeCluster returned:', JSON.stringify(resp, undefined, 2));
+    } catch (e) {
+      if (e.code === 'ResourceNotFoundException') {
+        console.log('received ResourceNotFoundException, this means the cluster has been deleted (or never existed)');
+        return { IsComplete: true };
+      }
+
+      console.log('describeCluster error:', e);
+      throw e;
+    }
+
+    return {
+      IsComplete: false
+    };
+  }
+
+  // ------
+  // UPDATE
+  // ------
+
+  public async onUpdate() {
+    const updates = analyzeUpdate(this.oldProps, this.newProps);
+
+    console.log(updates);
+
+    // if there is an update that requires replacement, go ahead and just create
+    // a new cluster with the new config. The old cluster will automatically be
+    // deleted by cloudformation upon success.
+    if (updates.replaceName || updates.replaceRole || updates.replaceVpc) {
+      return await this.onCreate();
+    }
```

Comment:
This is usually the behavior for replacement of resources with physical names. If the name is different, all is good. If the name is the same, it will simply fail creating the new cluster saying there's already a cluster with that name.

I guess we can proactively fail with a nicer error, but most CFN resources actually just fail with the error I described above, so I don't know if this is really required. Punting for now, but added some commentary.

Reply:
ok, I am convinced, added an explicit check with a nice error